src/CoreWCF.Primitives/src/CoreWCF/Security/LdapAdapter.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.DirectoryServices.Protocols;
+using System.Linq;
+using System.Text;
+using CoreWCF.IdentityModel.Claims;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Logging;
+
+namespace CoreWCF.Security
+{
+    internal static class LdapAdapter
+    {
+        public static List<Claim> RetrieveClaims(LdapSettings settings, string userName)
+        {
+            var user = userName;
+            var userAccountNameIndex = user.IndexOf('@');
+            var userAccountName = "";
+            if (userAccountNameIndex == -1)
+            {
+                userAccountNameIndex = user.IndexOf("\\");
+                if (userAccountNameIndex == -1)
+                    return null;
+                userAccountName = user.Substring(userAccountNameIndex+1);
+            }
+            else
+            {
+                userAccountName = user.Substring(0, userAccountNameIndex);
+            }
+
+            List<Claim> roleClaims = new List<IdentityModel.Claims.Claim>();
+            if (settings.ClaimsCache == null)
+            {
+                settings.ClaimsCache = new MemoryCache(new MemoryCacheOptions { SizeLimit = settings.ClaimsCacheSize });
+            }
```

Comment:
This code feels like it should live inside LdapSettings. The settings owns the ClaimsCacheSize, but isn't currently responsible for instantiating the MemoryCache with the correct size. It feels like the wrong ownership of responsibility when a different class can instantiate the cache with a limit different than settings specify.